### PR TITLE
[BREAKING] Deprecate keyword now

### LINKED ISSUE #4020: Deprecate "now" and just rely on "block.timestamp"
My reasoning is that `block.timestamp` is way more explicit in showing the intention, while `now` requires explanation: yes, `now` relates to the timestamp of the block controlled by the miner.

Warning in 0.6.x, error in 0.7.0.

### PR DESCRIPTION
**Pull request by hrkrshnn**
Fixes https://github.com/ethereum/solidity/issues/4020

TODO
1. [x] local variable `now` should work. ~~Remove the warning about shadowing.~~
2. [x] tests
3. [x] solidity-upgrade tool that replaces `now` with `block.timestamp`
4. [x] Upgrade external tests.

### CODE REVIEW

File: docs/contracts/constant-state-variables.rst [hunk 1]
```diff
@@ -44,7 +44,7 @@ Constant
 
 For ``constant`` variables, the value has to be a constant at compile time and it has to be
 assigned where the variable is declared. Any expression
-that accesses storage, blockchain data (e.g. ``now``, ``address(this).balance`` or
+that accesses storage, blockchain data (e.g. ``blockchain.timestamp``, ``address(this).balance`` or
```
Comment (axic) [inline docs/contracts/constant-state-variables.rst]: `block.timestamp`.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `81f816a63cc8593e93d3c6fced320f23c92f13a8`:
```
Coding style error:
 tools/solidityUpgrade/SourceTransform.h:283: << "\nNeeds to be fixed manually."

```
Please check that your changes are working as intended.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -2908,6 +2908,17 @@ bool TypeChecker::visit(Identifier const& _identifier)
 			);
 	}
 
+	if (MagicVariableDeclaration const* magicVar =
+			dynamic_cast<MagicVariableDeclaration const*>(annotation.referencedDeclaration))
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
	if (
		MagicVariableDeclaration const* magicVar =
		dynamic_cast<MagicVariableDeclaration const*>(annotation.referencedDeclaration)
	)
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -2908,6 +2908,17 @@ bool TypeChecker::visit(Identifier const& _identifier)
...
+	{
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: `{` not needed

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -2908,6 +2908,17 @@ bool TypeChecker::visit(Identifier const& _identifier)
...
+		if (magicVar->type()->category() == Type::Category::Integer)
+		{
+			solAssert(_identifier.name() == "now", "");
+			m_errorReporter
+				.typeError(_identifier.location(), "\"now\" has been deprecated. Use \"block.timestamp\" instead.");
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
			m_errorReporter.typeError(
				_identifier.location(),
				"\"now\" has been deprecated. Use \"block.timestamp\" instead."
			);
```

File: tools/solidityUpgrade/Upgrade070.cpp [hunk 5]
```diff
@@ -40,3 +40,22 @@ void DotSyntax::endVisit(FunctionCall const& _functionCall)
 			);
 	}
 }
+
+void NowKeyword::endVisit(Identifier const& _identifier)
+{
+	IdentifierAnnotation& annotation = _identifier.annotation();
+
+	if (MagicVariableDeclaration const* magicVar
```
Comment (chriseth) [inline tools/solidityUpgrade/Upgrade070.cpp]: style

File: tools/solidityUpgrade/SourceTransform.h [hunk 6]
```diff
@@ -267,6 +267,24 @@ class SourceTransform
 
 		return "";
 	}
+
+	static std::string nowUpdate(langutil::SourceLocation const& _location)
+	{
+		std::regex nowRegex{"now"};
+
+		if (regex_search(_location.text(), nowRegex))
+		{
```
Comment (chriseth) [inline tools/solidityUpgrade/SourceTransform.h]: not needed

File: Changelog.md [hunk 7]
```diff
@@ -3,6 +3,7 @@
 Breaking changes:
  * Type Checker: Disallow virtual for library functions.
  * Deprecated dot syntax for `value` and `gas`.
+ * Deprecated the identifier `now`.
```
Comment (chriseth) [inline Changelog.md:6]:
```suggestion
 * Deprecated the identifier `now` (use `block.timestamp`).
```

**Review by chriseth — APPROVED** [review]
Actually it's fine :)

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 8]
```diff
@@ -1025,29 +1025,6 @@ BOOST_AUTO_TEST_CASE(blockchain)
 	ABI_CHECK(callContractFunctionWithValue("someInfo()", 28), encodeArgs(28, u256("0x1212121212121212121212121212121212121212"), 7));
 }
 
-BOOST_AUTO_TEST_CASE(now)
```
Comment (axic) [inline test/libsolidity/SolidityEndToEndTest.cpp:1028]: Do we have a test like this somewhere now?

Comment (hrkrshnn) [reply test/libsolidity/SolidityEndToEndTest.cpp:1028]: @axic Good point. I'll add one back that doesn't use `now`.

**Pull request merged**